debug_module.tex

```diff
@@ -110,39 +110,21 @@ \section{Selecting Harts} \label{selectingharts}
 
 \section{Run Control} \label{runcontrol}
 
-The Debug Module can halt selected harts and allow them to run again using
-the \Rdmcontrol register.
-After selecting harts, it sets \Fhaltreq, then waits for \Fallhalted to
-indicate the harts are halted before clearing \Fhaltreq to 0. Setting \Fhaltreq
-has no effect on harts which are already halted.
-
-To resume, the debugger sets \Fresumereq. This action sets each selected hart's
-{\tt resumeack} bit to 0. Once a hart has resumed, it
-sets its {\tt resumeack} bit to 1. Thus, the
-debugger should wait for \Fallresumeack to indicate all
-harts have resumed before clearing \Fresumereq to 0.
-If a debugger reads \Fanyresumeack and \Fanyhalted in the same operation, then
-some selected harts must have resumed and then halted again.
-
-\begin{commentary}
-  When waiting for a hart to resume, a debugger should
-  examine \Fallresumeack, not \Fallrunning
-  or \Fallhalted,
-  because the hart may immediately halt again due to
-  trigger or step conditions.
-\end{commentary}
-
-Setting \Fresumereq on hart which is running
-clears {\tt resumeack} but
-may cause a hart which halts in the future to immediately
-resume. Debuggers should not depend on this behavior and should not
-set \Fresumereq for running harts.
-
-To halt or resume multiple harts at the same time, the debugger first
-sets the hart's bits in the hart array mask register, then
-follows the same procedure but with \Fhasel set to 1.
-Depending on the desired operation, the debugger might consider the {\tt any*}
-versions of the status instead of {\tt all*}.
+For every hart, the Debug Module contains 3 bits: halt request, resume request,
+and hart reset. (The hart reset bit is optional.) These bits all reset to 0. A
+debugger can write them for the currently selected harts through \Fhaltreq,
+\Fresumereq, and \Fhartreset in \Rdmcontrol.
+
+When a running hart sees a halt request, it responds by halting and asserting
+its halted signal. The halted signals of all selected harts are reflected in
```

Comment:
how about the halt summary registers?

It's a little strange to say above that "each hart has 3 bits" and now you are actually talking about a 4th. Did you consciously leave it off the list above? Or, maybe you should say above "there are 3 conceptual control bits", since below you go on to discuss even more bits/signals.

Reply:
I'm differentiating between bits (eg. FFs) and signals. Halt request is a bit set by the debugger, that persists even if a different hart is selected. Halted is a signal from a hart to the DM. It might be driven by a FF, or by some logic, etc. Is that confusing? How can I make it clear?

Reply:
I think it's not clear to me what you are trying to clarify :) The fact that the control signals are in flip flops or not is also just as much an implementation detail as the status signals.

I think you should just say "3 control signals" instead of "3 bits" and it would be clear to me.

Reply:
The control signals *have* to be in flip-flops, otherwise their value is lost when a different hart is selected.

The halt summary is probably not in flip-flops. It's probably just logic on the halted signals, which is sampled when the read happens.

Reply:
but all of this detail doesn't need to be in the spec. The fact that they are flops or not is really not necessary to clarify. Isn't a "halted signal" also probably in a flop? 

What is the higher level confusion we are trying to clear up?

Reply:
The part that has to be in the spec is that if a debugger sets haltreq on one hart, and then selects a different hart, that the first hart still sees a hartreq. There has to be state in the DM to track that.

Reply:
Not really. Maybe "haltreq" creates a message that gets sent by the Debug Module over some crazy network-on-chip, which responds with some packet later back to the Debug Module that tells the Debug Module it actually halted. The debug Module doesn't *have* to store the fact that it requested halt in a flop (unless we want to be able to read the haltreq bit when we switch back, which maybe we do, but that doesn't seem to be the thrust of this PR).

Reply:
But I think what is confusing in the spec and my example is that if a halt is requested, the hart halts,  then that hart gets a reset, then the hart should STILL be halted. So in my crazy NoC example, cores would need to know that they must stay halted as long as they don't get a "ok to not halt message" which is presumably which what an implementation would send when haltreq is written to 0.

Reply:
I agree that your example would be a valid implementation.

I've attempted to clarify that the bits are just state, and that there are additional signals from each hart to the DM.